source/common/router/router.cc

```diff
@@ -503,20 +503,40 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
   const auto& upstream_http_protocol_options = cluster_->upstreamHttpProtocolOptions();
 
   if (upstream_http_protocol_options.has_value()) {
-    const auto parsed_authority = Http::Utility::parseAuthority(headers.getHostValue());
-    if (!parsed_authority.is_ip_address_ && upstream_http_protocol_options.value().auto_sni()) {
-      callbacks_->streamInfo().filterState()->setData(
-          Network::UpstreamServerName::key(),
-          std::make_unique<Network::UpstreamServerName>(parsed_authority.host_),
-          StreamInfo::FilterState::StateType::Mutable);
+    absl::string_view sni_value;
+    bool should_set_sni = true;
+
+    // Check whether `alt_header_name` is specified.
+    const auto alt_header_name = upstream_http_protocol_options.value().alt_header_name();
+    if (alt_header_name.empty()) {
+      // Default the SNI value to the Host/Authority header.
+      const auto parsed_authority = Http::Utility::parseAuthority(headers.getHostValue());
+      should_set_sni = !parsed_authority.is_ip_address_;
+      sni_value = std::string(parsed_authority.host_);
+    } else {
+      // Use the header value from `alt_header_name` to set the SNI value.
+      const auto header_value = Http::HeaderUtility::getAllOfHeaderAsString(
+          headers, Http::LowerCaseString(alt_header_name));
+      if (header_value.result().has_value()) {
```

Comment:
just a question, is there anyone looking for a fallback to the Host/Authority header when alt_header's value is empty?

Reply:
Added logic to fallback to Host/Authority when the overridden header is not found or is empty. Also documented the same in the proto file.